src/Data/List/Relation/Unary/Sorted.agda

```diff
@@ -0,0 +1,141 @@
+        -- Split into 2 helper functions to make termination checker happy
+        merge-auxₗ : A → List A → List A → List A
+        merge-auxₗ x xs [] = x ∷ xs
+        merge-auxₗ x xs (y ∷ ys) with ≺-total x y
+        merge-auxₗ x xs (y ∷ ys) | inj₁ x≺y = x ∷ merge-auxᵣ xs y ys
+        merge-auxₗ x xs (y ∷ ys) | inj₂ y≺x = y ∷ merge-auxₗ x xs ys
+
+        merge-auxᵣ : List A → A → List A → List A
+        merge-auxᵣ [] y ys = y ∷ ys
+        merge-auxᵣ (x ∷ xs) y ys with ≺-total x y
+        merge-auxᵣ (x ∷ xs) y ys | inj₁ x≺y = x ∷ merge-auxᵣ xs y ys
+        merge-auxᵣ (x ∷ xs) y ys | inj₂ y≺x = y ∷ merge-auxₗ x xs ys
+
+      merge : List A → List A → List A
+      merge [] ys = ys
+      merge (x ∷ xs) ys = merge-auxₗ x xs ys
+
+      split : List A → List A × List A
+      split [] = [] , []
+      split (x ∷ xs) = x ∷ proj₂ r , proj₁ r where
+        r = split xs
+
+      split-length : ∀ xs → (let ys = split xs) (let n = length xs) → (length (proj₁ ys) ≡ ⌈ n /2⌉) × (length (proj₂ ys)) ≡ ⌊ n /2⌋
+      split-length [] = P.refl , P.refl
+      split-length (x ∷ []) = P.refl , P.refl
+      split-length (x₁ ∷ x₂ ∷ xs) = Σ.map (P.cong suc) (P.cong suc) (split-length xs)
+
+      split-length₁ : ∀ x₁ x₂ xs → (let ys = split (x₁ ∷ x₂ ∷ xs)) → length (proj₁ ys) < length (x₁ ∷ x₂ ∷ xs)
+      split-length₁ x₁ x₂ xs = begin
+        1 + length (proj₁ ys)           ≡⟨ P.cong (1 +_) (proj₁ ys-length) ⟩
+        1 + ⌈ length (x₁ ∷ x₂ ∷ xs) /2⌉ ≤⟨ ⌈n/2⌉<n (length xs) ⟩
+        length (x₁ ∷ x₂ ∷ xs)           ∎
+        where
+          open import Data.Nat
+          open import Data.Nat.Properties
+          open ≤-Reasoning
+          ys = split (x₁ ∷ x₂ ∷ xs)
+          ys-length = split-length (x₁ ∷ x₂ ∷ xs)
+
+      split-length₂ : ∀ x₁ x₂ xs → (let ys = split (x₁ ∷ x₂ ∷ xs)) → length (proj₂ ys) < length (x₁ ∷ x₂ ∷ xs)
+      split-length₂ x₁ x₂ xs = begin
+        1 + length (proj₂ ys)           ≡⟨ P.cong (1 +_) (proj₂ ys-length) ⟩
+        1 + ⌊ length (x₁ ∷ x₂ ∷ xs) /2⌋ ≤⟨ ⌊n/2⌋<n (length (x₂ ∷ xs)) ⟩
+        length (x₁ ∷ x₂ ∷ xs)           ∎
+        where
+          open import Data.Nat
+          open import Data.Nat.Properties
+          open ≤-Reasoning
+          ys = split (x₁ ∷ x₂ ∷ xs)
+          ys-length = split-length (x₁ ∷ x₂ ∷ xs)
+
+      merge-sort-aux : (xs : List A) → Acc _<_ (length xs) → List A
+      merge-sort-aux [] a = []
+      merge-sort-aux (x ∷ []) a = x ∷ []
+      merge-sort-aux (x₁ ∷ x₂ ∷ xs) (acc rs) = merge (merge-sort-aux (proj₁ ys) (rs (length (proj₁ ys)) lemma₁)) (merge-sort-aux (proj₂ ys) (rs (length (proj₂ ys)) lemma₂)) where
+        ys = split (x₁ ∷ x₂ ∷ xs)
+        ys-length = split-length (x₁ ∷ x₂ ∷ xs)
+        lemma₁ = split-length₁ x₁ x₂ xs
+        lemma₂ = split-length₂ x₁ x₂ xs
+
+      merge-sort : List A → List A
+      merge-sort xs = merge-sort-aux xs (<-wellFounded (length xs)) where
+        open import Data.Nat.Induction
+    open Merge using (merge; merge-sort) public
+
```

Comment:
This seems very nasty. I expect it should be a lot easier to implement
merge sort by going via `Data.Vec.Bounded`. I'll send a PR.

Reply:
On flip side it is much harder to prove that sort is permutation for vectors which is why it was easier for me to do it on lists.

(This is not included as I haven't intended to send the sort yet).